Added Version.h

Micky1979 opened the pull request:
request 1
Version.h contains copyright and update copyright year.
FileNVRAM_COPYRIGHT also include the init status printed via IOlog to
always have this usefull info in kernel log.

request 2
added boot arg (-NoFileNVRAM) to prevent the kext to load

Micky1979 committed:
Added Version.h

Version.h contains copyright and update copyright year.
FileNVRAM_COPYRIGHT also include the init status printed via IOlog to
always have this usefull info in kernel log.

Micky1979 committed:
Added -NoFileNVRAM boot arg

Using a boot arg to prevent the kext from starting. Usefull during
tests.

meklort commented on kext/FileNVRAM/Version.h:
```diff
@@ -0,0 +1,16 @@
+#ifndef FileNVRAM_Version_h
+#define FileNVRAM_Version_h
+
+#define FileNVRAM_VERSION "6.9.5"
```
For the Version.h change, please use something like the following for grabbing the version number (as a string) instead of hard coding the value:
            extern kmod_info_t kmod_info;
            kmod_info.version <-- this is where the version string is stored.

Micky1979 replied:
Hi meklort ok I'll do that way.

meklort commented on kext/FileNVRAM/FileNVRAM.cpp:
```diff
@@ -49,13 +51,26 @@ bool FileNVRAM::start(IOService *provider)
     
     //start is called upon wake for some reason.
-    if(mInitComplete)           return true;
+    if (mInitComplete) {
```
For consistency with existing code, please use the following style:
The { should always be on a line by itself.
After an if statement, there should be no space between the if and the (:
if()
instead of
if ()

meklort commented on kext/FileNVRAM/FileNVRAM.cpp:
```diff
@@ -49,13 +51,26 @@ bool FileNVRAM::start(IOService *provider)
     bool earlyInit = false;
     
-    LOG(NOTICE, "start() called (%d)\n", mInitComplete);
+    if (PE_parse_boot_argn("-NoFileNVRAM", gPEbuf, sizeof gPEbuf))
```
After an if statement, there should be no space between the if and the (:
if()
instead of
if ()

meklort commented on kext/FileNVRAM/FileNVRAM.cpp:
```diff
@@ -30,6 +31,7 @@ static IOPMPowerState sPowerStates[] = {
     {1, kIOPMPowerOn,  kIOPMPowerOn,  kIOPMPowerOn, 0, 0, 0, 0, 0, 0, 0, 0}
 };
 
+static char gPEbuf[256];
```
Is there any reason that this variable is global?

Micky1979 replied:
Well, no. I'll move it inside the start function.

Micky1979 committed:
Added boot arg to make the kext read only

-FileNVRAMro make the kext read only with no write operations. Usefull
for debugging if the IOBSD is not reachable.

Micky1979 committed:
readOnly to mReadOnly

as requested by the mantainer all variables get prefixed with low case
m char

Micky1979 committed:
using struct kmod_info to get the bundle version instead of hard coding it

using struct kmod_info to get the bundle version instead of hard coding
it as requested by the mantainer.
PE buffer moved inside the start function.

Micky1979 committed:
Fixed buffer lenght including null char

Fixed buffer lenght including null char calulating the size of
GUID+separator+key + null char using snprintf

meklort commented on kext/FileNVRAM/FileNVRAM.cpp:
```diff
@@ -217,7 +217,7 @@ void FileNVRAM::copyUnserialzedData(const char* prefix, OSDictionary* dict)
             if(prefix)
             {
-                size_t size = strlen(prefix) + sizeof(NVRAM_SEPERATOR) + strlen(key->getCStringNoCopy());
+                size_t size = strlen(prefix) + sizeof(NVRAM_SEPERATOR) + strlen(key->getCStringNoCopy()) +1;
```
This change is not needed, sizeof(NVRAM_SEPERATOR) includes the null character.

Micky1979 replied:
Something goes fixed this way, otherwise sometime(s) the key is not full and get truncated

Micky1979 replied:
I'll send you some screen shot of what I'm talking about

meklort replied:
OK, I'll take a look at the screenshots.
I'd suggest also testing with the line changed to the following instead:
size_t size = strlen(prefix) + strlen(NVRAM_SEPERATOR) + strlen(key->getCStringNoCopy()) +1;

This should be equivalent to the original line, and if it also shows an issue, then the bug you're trying to fix is caused by something else.

Micky1979 replied:
This is what happen:
`36C28AB5-6566-4C50-9EBD-CBB920F838§`
or
`D8F0CCF5-580E-4334-87B6-9FBBB831271D:NVRAFi`
Under /chosen/nvram all is ok, e.g is: `D8F0CCF5-580E-4334-87B6-9FBBB831271D:NVRAFile` and so on. 
Later I'll try to use strlen insead sizeoff as you suggested to see if is the same and let you know, although apparently we will have a char in less.
See you later

Micky1979 replied:
I confirm that `strlen(NVRAM_SEPERATOR) +1` works better, no more bug, committed on my fork.
Now I'm going to add a notification (using addMatchingNotification) to match IOMediaBSDClient instead of using the timeout waiting for IOBSD that is clearly not working at least in newer OSes. Don't know if this is supported by the old Tiger's kernel and I need to do some research before doing this (or doing this based on kernel version).
(Anyway now we have the Pike's safe read/write operations that allow the kext to never crash or doing damages. I prefear datas to be restored by the bootloader so that the read operation is just an optional never used and also the module should be fixed due to a bug in xml.c)

meklort replied:
Sounds good.

For reference:
- The kext was never meant to be used standalone. The reason for reading in the plist two-fold:
(1) As a backup in case if the bootloder (chameleon, or potentially clover if bootloader support is added) has an issue. The NVRAM contents really do need to be loaded in the bootloader, as certain processes *do* read the nvram the file-ystem is mounted.
(2) I often run the kext on real macs during development and to allow me to trap internal behavior of programs.

In any case,
Let's get this pull request merged in before you make the iobsd changes. I'll go though the latest changes and see if there's anything else.

Micky1979 replied:
Ok, waiting you Sir 👍

meklort commented on kext/FileNVRAM/FileNVRAM.cpp:
```diff
@@ -266,7 +266,7 @@ void FileNVRAM::copyEntryProperties(const char* prefix, IORegistryEntry* entry)
                 if(prefix)
                 {
-                    size_t size = strlen(prefix) + sizeof(NVRAM_SEPERATOR) + strlen(name);
+                    size_t size = strlen(prefix) + sizeof(NVRAM_SEPERATOR) + strlen(name) +1;
```
This change is not needed, sizeof(NVRAM_SEPERATOR) includes the null character.

meklort commented on kext/FileNVRAM/FileNVRAM.cpp:
```diff
@@ -305,7 +305,7 @@ void FileNVRAM::copyEntryProperties(const char* prefix, IORegistryEntry* entry)
             if(prefix)
             {
-                size_t size = strlen(prefix) + sizeof(NVRAM_SEPERATOR) + strlen(key->getCStringNoCopy());
+                size_t size = strlen(prefix) + sizeof(NVRAM_SEPERATOR) + strlen(key->getCStringNoCopy()) +1;
```
This change is not needed, sizeof(NVRAM_SEPERATOR) includes the null character.

meklort commented on kext/FileNVRAM/FileNVRAM.cpp:
```diff
@@ -47,15 +49,36 @@ void FileNVRAM::setPath(OSString* path)
     bool earlyInit = false;
     
-    LOG(NOTICE, "start() called (%d)\n", mInitComplete);
+    if (PE_parse_boot_argn("-NoFileNVRAM", gPEbuf, sizeof gPEbuf))
```
for consistency, change to:
if()
{
instead of:
if ()
{

Micky1979 replied:
done, also extended this to all the code

meklort commented on kext/FileNVRAM/FileNVRAM.cpp:
```diff
@@ -47,15 +49,36 @@ void FileNVRAM::setPath(OSString* path)
+        return false; /* following module disable method */
+    }
+    
+    if (PE_parse_boot_argn("-FileNVRAMro", gPEbuf, sizeof gPEbuf)) /* read only */
```
for consistency, change to:
if()
{
instead of:
if ()
{

Micky1979 replied:
done

meklort commented on kext/FileNVRAM/FileNVRAM.cpp:
```diff
@@ -47,15 +49,36 @@ void FileNVRAM::setPath(OSString* path)
     
     //start is called upon wake for some reason.
-    if(mInitComplete)           return true;
+    if (mInitComplete) {
```
for consistency, change to:
if()
{
instead of:
if () {

meklort commented on kext/FileNVRAM/FileNVRAM.cpp:
```diff
@@ -91,6 +114,15 @@ bool FileNVRAM::start(IOService *provider)
         bootnvram->detachFromParent(root, gIODTPlane);
     }
+    
+    if (mReadOnly) {
```
for consistency, change to:
if()
{
instead of:
if () {

meklort commented on kext/FileNVRAM/FileNVRAM.cpp:
```diff
@@ -333,8 +365,10 @@ void FileNVRAM::doSync(void)
     
-    if(!mFilePath) return;
-    if(!mSafeToSync) return;
+    if (!mFilePath || !mSafeToSync || mReadOnly) {
```
for consistency, change to:
if()
{
instead of:
if () {

meklort commented on kext/FileNVRAM/FileNVRAM.cpp:
```diff
@@ -47,15 +49,36 @@ void FileNVRAM::setPath(OSString* path)
 
 bool FileNVRAM::start(IOService *provider)
 {
+    char gPEbuf[256];
```
remove g from name, variable is not global.

Micky1979 replied:
done

Micky1979 committed:
Safe read/write operation

Safe read/write operation imported from Pike R (FileIO.h/c removed).
Alpha. Code formatting uniformed.

Micky1979 committed:
fixed typo

fixed typo importing changes

Micky1979 committed:
Using strlen instead of sizeof

Using strlen (+1) instead of sizeof calculating buffer size.

meklort commented on kext/FileNVRAM/FileNVRAM.h:
```diff
@@ -154,7 +154,10 @@ class FileNVRAM : public IODTNVRAM
     
     static IOReturn dispatchCommand( OSObject* owner, void* arg0, void* arg1, void* arg2, void* arg3 );
     
+    virtual IOReturn write_buffer(char* buffer, vfs_context_t ctx);
```
mCtx is a member variable. the ctx parameter can be removed now that it's part of the class.

Micky1979 replied:
roger, il do it

meklort commented on kext/FileNVRAM/FileNVRAM.h:
```diff
@@ -154,7 +154,10 @@ class FileNVRAM : public IODTNVRAM
     static IOReturn dispatchCommand( OSObject* owner, void* arg0, void* arg1, void* arg2, void* arg3 );
     
+    virtual IOReturn write_buffer(char* buffer, vfs_context_t ctx);
+    virtual IOReturn read_buffer(char** buffer, uint64_t* length, vfs_context_t ctx);
```
mCtx is a member variable. the ctx parameter can be removed now that it's part of the class.

meklort commented on kext/FileNVRAM/FileNVRAM.cpp:
```diff
@@ -47,15 +49,38 @@ void FileNVRAM::setPath(OSString* path)
     bool earlyInit = false;
     
-    LOG(NOTICE, "start() called (%d)\n", mInitComplete);
+    if(PE_parse_boot_argn("-NoFileNVRAM", PEbuf, sizeof peBuf))
```
I'd suggest moving the "-NoFileNVRAM" and "-FileNVRAMro" strings to the header file so all of the options are in one spot.

meklort commented:
What's the purpose of the read only flag?
Is it to:
- Make the NVRAM read only?
- Make the NVRAM non-persistant?

From the changes, it looks like it doesn't actually make the nvram read only, what it does it make it non-persistant, the old nvram contents will be restores on next reboot. Is this what you want?

meklort commented:
Would you also please update the NEWS and README in the docs folder with:
- A description of the changes (in NEWS, you can add a new section for this, but for the version number I'd put something like "develop" that indicates it's not released officially yet.)
- The usage of the new boot arguments (in README, under usage).

After that, I think it's OK to merge.

Micky1979 committed:
typo

typo importing changes

Micky1979 committed:
Fixed compilation, added default nvram path macro

Fixed compilation, added default nvram path macro. Boot key are now
macros as suggested by the mantainer

Micky1979 committed:
Added debug boot key

Added debug boot key to enable debugging from boot args

Micky1979 committed:
restored old default file path

as suggested by the mantainer. Anyway this should be changed soon since
chameleon move to a direction of using one Extra folder, that can be
inside in the ESP for example.

Micky1979 commented:
Hi meklort, sorry it seem that I cannot reply to all of your question under them. 
"read only" can allow clover to dump the nvram using its existing system. Mind that macOS with Clover (in the case) can be installed on logical volumes, so the nvram should be saved to the boot partition (the ESP), but this cannot be done by this kext.

meklort commented:
Ok, looks like I see two remaining changes needed before merge:
(1) FileIO.c/h should be removed from the xcode  project. If you want, I can do that.
(2) Documentation update in the docs folder.

Micky1979 committed:
Update README and NEWS

Micky1979 commented:
Yeas please, update the Xcode project, I don't want to add more garbage. The docs are updated.

Micky1979 commented:
Thanks, tomorrow will continue with other things.

meklort commented:
Sounds good. I'll try to setup Travis-CI so that future commits and pull requests auto-build as well.
